src/client/datascience/data-viewing/data-wrangler/dataWrangler.ts

```diff
@@ -0,0 +1,685 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+import '../../../common/extensions';
+
+import { inject, injectable, named } from 'inversify';
+import * as path from 'path';
+import * as uuid from 'uuid/v4';
+import * as fsextra from 'fs-extra';
+import {
+    Disposable,
+    EventEmitter,
+    Memento,
+    notebooks as vscNotebook,
+    NotebookCell,
+    NotebookCellExecutionState,
+    NotebookCellExecutionStateChangeEvent,
+    ViewColumn,
+    WebviewPanel
+} from 'vscode';
+
+import {
+    IApplicationShell,
+    ICommandManager,
+    IWebviewPanelProvider,
+    IWorkspaceService,
+    IDocumentManager
+} from '../../../common/application/types';
+import { EXTENSION_ROOT_DIR, PYTHON_LANGUAGE, UseCustomEditorApi } from '../../../common/constants';
+import { traceError, traceInfo, traceWarning } from '../../../common/logger';
+import { GLOBAL_MEMENTO, IConfigurationService, IDisposable, IMemento, Resource } from '../../../common/types';
+import * as localize from '../../../common/utils/localize';
+import { noop } from '../../../common/utils/misc';
+import { Commands, HelpLinks, Identifiers } from '../../constants';
+import { JupyterDataRateLimitError } from '../../jupyter/jupyterDataRateLimitError';
+import {
+    ICodeCssGenerator,
+    IJupyterVariables,
+    INotebookEditorProvider,
+    IThemeFinder,
+    WebViewViewChangeEventArgs
+} from '../../types';
+import { WebviewPanelHost } from '../../webviews/webviewPanelHost';
+import { isValidSliceExpression, preselectedSliceExpression } from '../../../../datascience-ui/data-explorer/helpers';
+import { addNewCellAfter, updateCellCode } from '../../notebook/helpers/executionHelpers';
+import { InteractiveWindowMessages } from '../../interactive-common/interactiveWindowTypes';
+import { serializeLanguageConfiguration } from '../../interactive-common/serialization';
+import { CssMessages } from '../../messages';
+import { IDataFrameInfo, IGetRowsRequest, IGetSliceRequest } from '../types';
+import { DataWranglerMessageListener } from './dataWranglerMessageListener';
+import {
+    IDataWranglerMapping,
+    IDataWrangler,
+    IDataWranglerDataProvider,
+    DataWranglerMessages,
+    DataWranglerCommands,
+    IRenameColumnsRequest,
+    IHistoryItem,
+    IDropRequest,
+    INormalizeColumnRequest,
+    IFillNaRequest,
+    IDropDuplicatesRequest,
+    IDropNaRequest,
+    OpenDataWranglerSetting,
+    IDataWranglerJupyterVariableDataProvider,
+    IDataWranglerJupyterVariableDataProviderFactory
+} from './types';
+
+const PREFERRED_VIEWGROUP = 'JupyterDataWranglerPreferredViewColumn';
+const dataWranglerDir = path.join(EXTENSION_ROOT_DIR, 'out', 'datascience-ui', 'viewers');
+@injectable()
+export class DataWrangler extends WebviewPanelHost<IDataWranglerMapping> implements IDataWrangler, IDisposable {
```

Comment:
I'm wondering if this class and the DataViewer can share anything? They are pretty similar.

Reply:
They are definitely very similar. So are some of the other classes, I really really want to figure out the best way to re-use code.  Here are my thoughts: abstract classes, but it won't work because the classes `extend` the same class but the `Mapping` is different. Same for the React classes with the props. Unless I made it so they used the same mappings, and made the data wrangler ones optional. I could also have a dataViewingHelper (probably needs a better name) class that both dataViewer and dataWrangler call when the code is the same but there will still probably be a bit of duplicate code. Any other ideas or thoughts on the above? They are very welcomed and thank you for your review. I appreciate it a lot.

Reply:
DataWrangler could probably be a subclass of DataViewer on the extension side? Its handleMessage might have some extra message handlers but otherwise they seem identical. 

On the react side, the message handling could be abstracted into a base class for both (and others) or you could use some object to handle the messages.  Actually I think the tmlanguage and onigasm stuff is already abstracted using redux, but maybe the datawrangler/dataviewer don't bother with redux. If they don't, then just creating a common base class or some other object to handle those messages seems like it would work.

Reply:
Alright, so I did the `extends` implementation
I decided to just add DataWranglerMessages to IDataViewerMapping
I played around with the constructor because I wasn't sure about the best way to get the different values into WebviewPanelHost<IDataViewerMapping> (Eg. title, directory strings). What do you think?

Reply:
Couldn't DataWrangler extend DataViewer? They're very similar

Reply:
Yep, that's what I did in my most recent commits.